### PR TITLE
MQTT codec v1.1

### PR DESCRIPTION
**Pull request by nayato**
- reworked decoder to use remaining length as an indicator for how many bytes need to be buffered before it makes sense to parse anything beyond the remaining length field in the packet.
- added happy path tests for re-coding for all packet types (#6). 
- bug fixes through testing.

### CODE REVIEW

File: src/DotNetty.Codecs.Mqtt/MqttDecoder.cs [hunk 1]
```diff
@@ -328,25 +275,19 @@ void DecodeConnectVariableHeader(IByteBuffer buffer)
...
-            //    throw new DecoderException(string.Format("Unexpected protocol level. Expected: {0}. Actual: {1}", ProtocolLevel, protocolLevel));
-            //}
+            if (packet.ProtocolVersion != Util.ProtocolVersion)
+            {
+                MqttEventSource.Log.Error(string.Format("Unexpected protocol level. Expected: {0}. Actual: {1}", Util.ProtocolVersion, packet.ProtocolVersion), string.Empty);
+                var connAckPacket = new ConnAckPacket();
+                connAckPacket.ReturnCode = ConnectReturnCode.RefusedUnacceptableProtocolVersion;
+                context.WriteAndFlushAsync(connAckPacket)
+                    .ContinueWith((t, ctx) => ((IChannelHandlerContext)ctx).CloseAsync(), context);
+            }
```
Comment (manusr) [inline src/DotNetty.Codecs.Mqtt/MqttDecoder.cs]: let's throw DecoderException here after writing + flushing CONNACK

File: src/DotNetty.Codecs.Mqtt/MqttDecoder.cs [hunk 2]
```diff
@@ -97,15 +89,16 @@ void DecodeFixedHeader(IByteBuffer buffer)
 
             int signature = buffer.ReadByte();
 
-            BufferReadResult<int> remainingLengthResult = this.DecodeRemainingLength(buffer);
-            if (!remainingLengthResult.IsSuccessful)
+            if (!this.TryDecodeRemainingLength(buffer, out this.remainingLength))
```
Comment (manusr) [inline src/DotNetty.Codecs.Mqtt/MqttDecoder.cs]: Let's put the RequestReplay here as well for consistency

File: src/DotNetty.Codecs.Mqtt/MqttDecoder.cs [hunk 3]
```diff
@@ -15,8 +15,6 @@ public sealed class MqttDecoder : ReplayingDecoder<MqttDecoder.ParseState>
         public enum ParseState
         {
             FixedHeader,
-            VariableHeader,
-            Payload,
             BadMessage
         }
```
Comment (manusr) [inline src/DotNetty.Codecs.Mqtt/MqttDecoder.cs]: remainingLength and currentPacket should no longer be instance members

File: src/DotNetty.Codecs.Mqtt/MqttDecoder.cs [hunk 4]
```diff
@@ -39,13 +37,7 @@ protected override void Decode(IChannelHandlerContext context, IByteBuffer input
                 switch (this.State)
                 {
                     case ParseState.FixedHeader:
-                        this.DecodeFixedHeader(input);
```
Comment (manusr) [inline src/DotNetty.Codecs.Mqtt/MqttDecoder.cs]: add rest of code after switch statement to this case statement

File: src/DotNetty.Codecs.Mqtt/MqttDecoder.cs [hunk 5]
```diff
@@ -447,56 +369,36 @@ void DecodePacketIdVariableHeader(IByteBuffer buffer, PacketWithId packet, bool
...
-                throw new DecoderException(string.Format("Remaining Length is not big enough to accomodate at least 2 bytes (for packet identifier). Available value: {0}",
-                    this.remainingLength));
-            }
-
-            BufferReadResult<int> packetIdResult = this.DecodeShort(buffer);
-            if (packetIdResult.IsSuccessful)
-            {
-                Util.ValidatePacketId(packetIdResult.Value);
-            }
+            BufferReadResult<int> packetIdResult = this.DecodeShort(buffer, currentRemainingLength);
```
Comment (manusr) [inline src/DotNetty.Codecs.Mqtt/MqttDecoder.cs]: return result directly

File: src/DotNetty.Codecs.Mqtt/MqttDecoder.cs [hunk 6]
```diff
@@ -447,56 +369,36 @@ void DecodePacketIdVariableHeader(IByteBuffer buffer, PacketWithId packet, bool
...
-                packet.WillTopic = willTopicResult.Value;
-                BufferReadResult<int> willMessageLengthResult = this.DecodeShort(buffer);
-                if (!willMessageLengthResult.IsSuccessful)
-                {
-                    return;
-                }
+                packet.WillTopicName = willTopicResult.Value;
+                BufferReadResult<int> willMessageLengthResult = this.DecodeShort(buffer, this.remainingLength - bytesConsumed);
                 int willMessageLength = willMessageLengthResult.Value;
                 if (!buffer.IsReadable(willMessageLength))
```
Comment (manusr) [inline src/DotNetty.Codecs.Mqtt/MqttDecoder.cs]: remove RequestReplay() in this entire method. In fact, any RequestReplay() after the ensure readable is pointless

File: src/DotNetty.Codecs.Mqtt/MqttDecoder.cs [hunk 7]
```diff
@@ -722,26 +596,35 @@ BufferReadResult<string> DecodeString(IByteBuffer buffer, int currentRemainingLe
...
         }
 
-        struct BufferReadResult<T>
+        bool EnsureCompletePacketIsReadable(IByteBuffer buffer)
         {
-            public readonly T Value;
-            public readonly int BytesConsumed;
+            if (!buffer.IsReadable(this.remainingLength))
+            {
+                this.RequestReplay();
```
Comment (manusr) [inline src/DotNetty.Codecs.Mqtt/MqttDecoder.cs]: caller should request replay?

**Comment by nayato** [comment]
@manusr please check again.

**Comment by manusr** [comment]
Looks good to me

**Pull request merged**